### PR TITLE
test(azure_policy): add foundation test cases

### PR DESCRIPTION
**Pull request by anakrish**
YAML-driven test cases for the core Azure Policy compiler. These cover alias resolution, field conditions, logical operators, type coercion, count expressions, template functions, effect compilation, and policy definition parsing.

24 files, each a self-contained scenario exercised by the test runner in the companion code PR.

**Part 1 of 3** — split from #695 to enable Copilot review on each PR individually.

### CODE REVIEW

File: tests/azure_policy/cases/value_conditions.yaml [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    want_effect: "deny"
+
+  - note: value_notIn_array
+    policy_rule: |
+      {
+        "if": {
+          "value": "[parameters('location')]",
+          "notIn": ["eastus", "westus"]
+        },
+        "then": { "effect": "deny" }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/value_conditions.yaml]: Good catch — renamed to `value_not_in_array` for consistent snake_case.

File: tests/azure_policy/cases/type_coercion.yaml [hunk 2]
```diff
@@ -0,0 +1,358 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Type Coercion Test Suite
+# Tests string↔number coercion, null/undefined handling, and other
+# implicit type behaviors documented in semantic-behaviors.md.
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/type_coercion.yaml]: Fixed — removed the dangling reference.

File: tests/azure_policy/cases/type_coercion.yaml [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+          "field": "properties.enabled",
+          "equals": true
+        },
+        "then": { "effect": "audit" }
+      }
+    resource:
+      properties:
+        enabled: "true"
+    want_effect: "audit"
+
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/type_coercion.yaml]: Agreed, the old name was misleading. Renamed to `bool_string_coercion_matches` since the test verifies that string↔bool coercion causes the condition to match.

File: tests/azure_policy/cases/implicit_allof.yaml [hunk 4]
```diff
@@ -0,0 +1,554 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Implicit allOf Test Suite
+# When a field with [*] appears in a condition outside count, Azure Policy
+# applies implicit allOf semantics: ALL elements must satisfy the operator.
+# Empty or missing arrays evaluate to true (vacuous truth).
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/implicit_allof.yaml]: Good find — the header was wrong. Updated to distinguish empty arrays (vacuous truth → true) from missing arrays (not iterable → false).

File: tests/azure_policy/cases/parse_errors.yaml [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+      {
+        "if": {
+          "value": "[concat()]",
+          "equals": ""
+        },
+        "then": { "effect": "deny" }
+      }
+    resource:
+      type: "any"
+    want_effect: "deny"
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/parse_errors.yaml:269]: This case is intentionally not a parse error — `concat()` with zero args is valid and returns empty string. Renamed the section header to "Edge cases: expressions that look malformed but are actually valid" and the note to `concat_zero_args_is_valid` to make the intent clear.

File: tests/azure_policy/cases/parse_errors.yaml [hunk 6]
```diff
@@ -0,0 +1,269 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Parse Error Test Suite
+# Tests that malformed policy JSON and invalid constructs are properly rejected.
+# These test cases are expected to fail parsing.
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/parse_errors.yaml:6]: Good point — updated the header to reflect that the file is mostly parse-error cases but also includes valid edge cases.

File: tests/azure_policy/cases/template_functions_extra.yaml [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+    resource:
+      type: "any"
+    want_effect: "audit"
+
+  - note: fn_base64ToJson
+    policy_rule: |
+      {
+        "if": {
+          "value": "[base64ToJson('eyJrZXkiOiJ2YWx1ZSJ9')]",
+          "notEquals": ""
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/template_functions_extra.yaml]: Nice catch — the old assertion was too weak. Changed to access `.key` on the decoded object and assert `equals: "value"`, which validates actual decoding.

File: tests/azure_policy/cases/fields.yaml [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+      properties:
+        networkAcls:
+          default-action: "Allow"
+    want_effect: "deny"
+
+  - note: field_bracket_notation_double_quote
+    policy_rule: |
+      {
+        "if": {
+          "field": "properties.settings['log-level']",
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/fields.yaml:259]: Good catch — changed to escaped double quotes so this case actually tests a distinct bracket notation style.

File: tests/azure_policy/cases/template_functions_extra.yaml [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+    resource:
+      type: "any"
+    want_effect: "audit"
+
+  - note: fn_float_from_string
+    policy_rule: |
+      {
+        "if": {
+          "value": "[float('3.14')]",
+          "greaterOrEquals": "3.14"
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/template_functions_extra.yaml]: Agreed — changed to `equals: 3.14` against a numeric RHS to properly validate the return type and value.

File: tests/azure_policy/cases/implicit_allof.yaml [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+          "equals": "active"
+        },
+        "then": { "effect": "deny" }
+      }
+    resource:
+      items:
+        - { "status": "active" }
+        - { "status": "inactive" }
+        - { "status": "active" }
+    want_no_effect: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/implicit_allof.yaml]: Good find — `want_no_effect` wasn't even handled by the test runner, so these assertions were silently passing. Replaced all 15 occurrences with `want_undefined: true` across both files.

File: tests/azure_policy/cases/field_wildcard_collect.yaml [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+          "equals": 5
+        },
+        "then": { "effect": "deny" }
+      }
+    resource:
+      items:
+        - "a"
+        - "b"
+        - "c"
+    want_no_effect: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/field_wildcard_collect.yaml]: Fixed — same issue as above. All `want_no_effect` replaced with `want_undefined: true`.

File: tests/azure_policy/cases/template_functions_extra.yaml [hunk 12]
```diff
@@ -0,0 +1,561 @@
...
+      type: "any"
+    want_effect: "audit"
+
+  - note: fn_dataUri
+    policy_rule: |
+      {
+        "if": {
+          "value": "[dataUri('Hello')]",
+          "contains": "base64,"
+        },
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/template_functions_extra.yaml:190]: This is actually valid JSON — the comma is inside the string value `"base64,"` (data URIs use `base64,` as a delimiter before the encoded payload). Line 190 shows `},` which is the comma separating the `"if"` and `"then"` keys in the outer object. The test passes and correctly validates `dataUri()` output.

File: tests/azure_policy/cases/field_wildcard_collect.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/azure_policy/cases/field_wildcard_collect.yaml]: Fixed — this was supposed to be replaced in our earlier pass but was missed when propagating changes from the working branch to the split branch. All 15 occurrences of `want_no_effect` are now replaced with `want_undefined: true` across both `implicit_allof.yaml` and `field_wildcard_collect.yaml`. Also adding `#[serde(deny_unknown_fields)]` to the TestCase struct in PR #700 so typos and unknown fields cause immediate deserialization errors rather than being silently ignored.

**Review by dpokluda — APPROVED** [review]
